motionplan/collision.go

```diff
@@ -55,169 +54,110 @@ type collisionEntity struct {
+	reportDistances bool
 }
 
-func (sce spaceCollisionEntities) checkCollision(key, test *collisionEntity, reportDistances bool) (float64, error) {
-	encompassed, err := key.geometry.EncompassedBy(test.geometry)
-	if err != nil {
-		return math.NaN(), err
+func newEntityGraph(x, y *collisionEntities, reportDistances bool) *entityGraph {
+	distances := make([][]float64, x.count())
+	for i := range distances {
+		distances[i] = make([]float64, y.count())
 	}
-	// TODO(rb): EncompassedBy should also report distance required to resolve the collision
-	if !encompassed {
-		return 1, nil
+	return &entityGraph{
+		x:               x,
+		y:               y,
+		distances:       distances,
+		reportDistances: reportDistances,
+		triangular:      x == y,
 	}
-	return -1, nil
 }
 
-func (sce spaceCollisionEntities) reportCollisions(distances []float64) []int {
-	collisionIndices := make([]int, 0)
-	for i := range distances {
-		if distances[i] >= spatial.CollisionBuffer {
-			collisionIndices = append(collisionIndices, i)
-		} else {
-			return []int{}
-		}
-	}
-	return collisionIndices
+func (cg *entityGraph) getIndices(xName, yName string) (int, int, bool) {
+	i := cg.x.indexFromName(xName)
+	j := cg.y.indexFromName(yName)
+	return i, j, i >= 0 && j >= 0
 }
 
-// collisionGraph is an implementation of an undirected graph used to track collisions between two set of CollisionEntities.
 type collisionGraph struct {
-	// key CollisionEntities
-	key *ObjectCollisionEntities
-
-	// test CollisionEntities are the set of CollisionEntities from which the collisionGraph takes its
-	// collisionCheckFn and collisionReportFn functions to check and report collisions between the
-	// test CollisionEntities and key Collision Entities
-	test CollisionEntities
-
-	// adjacencies is 2D array encoding edges between collisionEntiies in the collisionGraph.
-	// if adjacencies[i][j] >= 0 this corresponds to an edge between the entities at key[i] and test[j]
-	adjacencies [][]float64
-
-	// triangular is a bool that describes if the adjacencies matrix is triangular, which will be the case when key == test
-	triangular bool
-
-	reportDistances bool
+	entityGraph
 }
 
 // newCollisionGraph instantiates a collisionGraph object and checks for collisions between the key and test sets of CollisionEntities
 // collisions that are reported in the reference CollisionSystem argument will be ignore and not stored as edges in the graph.
-func newCollisionGraph(
-	key *ObjectCollisionEntities,
-	test CollisionEntities,
-	reference *CollisionSystem,
-	reportDistances bool,
-) (*collisionGraph, error) {
-	var err error
-	cg := &collisionGraph{
-		key:             key,
-		test:            test,
-		adjacencies:     make([][]float64, key.count()),
-		triangular:      key == test,
-		reportDistances: reportDistances,
+func newCollisionGraph(x, y *collisionEntities, reference *collisionGraph, reportDistances bool) (cg *collisionGraph, err error) {
+	if y == nil {
+		y = x
```

Comment:
I like this shortcut, but are there any ways we can use this to improve performance even further when checking the `i`th and `j`th ranges, since those entities should be the same in both `collisionEntities*`?

Reply:
I don't think this would be a big speedup if a speedup at all.  You would have to add logic to check if this condition applies and that would probably cancel out any performance gains.  The operation to get the index is a simple array index, so should be incredibly cheap.  

This does have me thinking though, I could split up the triangular/non triangular graphs into different collisionGraphs, this might make things a tiny bit faster at the cost of rewriting a lot of code